hydra-chain-observer/src/Hydra/ChainObserver.hs

```diff
@@ -162,32 +169,38 @@ chainSyncClient tracer networkId startingPoint observerHandler =
                 BlockInMode ConwayEra (Block _header conwayTxs) -> mapMaybe convertTx conwayTxs
                 BlockInMode BabbageEra (Block _header babbageTxs) -> babbageTxs
                 _ -> []
+
+              (BlockInMode _ (Block (BlockHeader _ _ blockNo) _)) = blockInMode
+              point = chainTipToChainPoint tip
           traceWith
             tracer
             RollForward
-              { point = chainTipToChainPoint tip
+              { point
               , receivedTxIds = getTxId . getTxBody <$> txs
               }
           let (utxo', observations) = observeAll networkId utxo txs
-          -- FIXME we should be exposing OnChainTx instead of working around NoHeadTx.
-          forM_ observations (maybe (pure ()) (traceWith tracer) . logObservation)
-          observerHandler observations
+              onChainTxs = mapMaybe convertObservation observations
+          forM_ onChainTxs (traceWith tracer . logOnChainTx)
+          let observationsAt =
+                fmap convertObservation observations <&> \case
+                  Just onChainTx -> HeadObservationAt point blockNo (Just onChainTx)
+                  Nothing -> HeadObservationAt point blockNo Nothing
```

Comment:
Should: model this better

I would say that this is not really a `HeadObservationAt` if it observes `Nothing`, right?

Instead of having a wrapped `Maybe`, you could have a sum type where one is just a `Tick` and another is a `Observation`. You could consider re-using the `ChainEvent` type, which is basically just that!

Reply:
re-using the `ChainEvent` type its different from what I need.

however I refactored the model to follow the same approach.